File: recursion/javascript/solutions/tails.js
/**
 * Solution for tails.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');

function tails(list) {
  if (isEmpty(list)) {
    return prepend(EMPTY_LIST, EMPTY_LIST);
  }

  let [first, rest] = unprepend(list);

  return prepend(list, tails(rest));
}

if (require.main === module) {
  let result = tails(Node.fromArray([1, 2, 3]));
  let tailsList = toArray(result).map((t) => (t === null ? [] : toArray(t)));
  console.assert(JSON.stringify(tailsList) === '[[1,2,3],[2,3],[3],[]]');

  let emptyResult = tails(EMPTY_LIST);
  let emptyTails = toArray(emptyResult).map((t) => (t === null ? [] : toArray(t)));
  console.assert(JSON.stringify(emptyTails) === '[[]]');

  console.log('All tests passed!');
}

module.exports = { tails };

File: recursion/python/solutions/tails.py
"""Solution for tails."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list


def tails(lst):
    """
    Return list of all suffixes (tails).

    tails(1 -> 2 -> 3) => (1 -> 2 -> 3) -> (2 -> 3) -> (3) -> ()
    tails(()) => ()
    """
    if is_empty(lst):
        return prepend(EMPTY_LIST, EMPTY_LIST)

    first, rest = unprepend(lst)

    return prepend(lst, tails(rest))


if __name__ == '__main__':
    # Test cases
    result = tails(Node.from_list([1, 2, 3]))
    result_list = to_python_list(result)
    assert to_python_list(result_list[0]) == [1, 2, 3]
    assert to_python_list(result_list[1]) == [2, 3]
    assert to_python_list(result_list[2]) == [3]
    assert to_python_list(result_list[3]) == []

    result = tails(EMPTY_LIST)
    result_list = to_python_list(result)
    assert len(result_list) == 1
    assert to_python_list(result_list[0]) == []

    print("All tests passed!")
